src/Integration.Vsix/SonarLintTracker/IssueTracker.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            var oldMarkers = this.Factory.CurrentSnapshot;
+            var newMarkers = new IssuesSnapshot(this.FilePath, oldMarkers.VersionNumber + 1);
+
+            // Copy all of the old errors to the new errors unless the error was affected by the text change
+            foreach (var marker in oldMarkers.IssueMarkers)
+            {
+                var newMarker = IssueMarker.CloneAndTranslateTo(marker, currentSnapshot);
+                if (newMarker != null)
+                {
+                    Debug.Assert(newMarker.Span.Length == marker.Span.Length);
+                    newMarkers.IssueMarkers.Add(newMarker);
+                }
+            }
+
+            return newMarkers;
+        }
+
+        internal void UpdateIssues(IList<Issue> issues)
+        {
+            var oldSnapshot = this.Factory.CurrentSnapshot;
+            var newSnapshot = new IssuesSnapshot(this.FilePath, oldSnapshot.VersionNumber + 1);
+
+            foreach (Issue issue in issues)
+            {
+                int startPos = currentSnapshot.GetLineFromLineNumber(issue.StartLine - 1).Start.Position + issue.StartLineOffset;
+                var start = new SnapshotPoint(currentSnapshot, startPos);
+
+                int endPos = currentSnapshot.GetLineFromLineNumber(issue.EndLine - 1).Start.Position + issue.EndLineOffset;
+                var end = new SnapshotPoint(currentSnapshot, endPos);
+
+                newSnapshot.IssueMarkers.Add(new IssueMarker(issue, new SnapshotSpan(start, end)));
+            }
+
+            SnapToNewSnapshot(newSnapshot);
+        }
+
+        private void SnapToNewSnapshot(IssuesSnapshot snapshot)
+        {
+            // Tell our factory to snap to a new snapshot.
+            this.Factory.UpdateMarkers(snapshot);
+
+            // Tell the provider to mark all the sinks dirty (so, as a side-effect, they will start an update pass that will get the new snapshot
+            // from the factory).
+            provider.UpdateAllSinks();
+
+            UpdateMarkers(currentSnapshot, snapshot);
+
+            this.snapshot = snapshot;
+        }
+
+        internal void UpdateMarkers(ITextSnapshot currentSnapshot, IssuesSnapshot snapshot)
+        {
+            var oldSnapshot = this.snapshot;
+
+            var h = this.TagsChanged;
+            if (h != null)
+            {
+                // Raise a single tags changed event over the span that could have been affected by the change in the errors.
+                int start = int.MaxValue;
+                int end = int.MinValue;
+
+                if (oldSnapshot != null && oldSnapshot.IssueMarkers.Count > 0)
+                {
+                    start = oldSnapshot.IssueMarkers[0].Span.Start.TranslateTo(currentSnapshot, PointTrackingMode.Negative);
+                    end = oldSnapshot.IssueMarkers[oldSnapshot.IssueMarkers.Count - 1].Span.End.TranslateTo(currentSnapshot, PointTrackingMode.Positive);
+                }
```

Comment:
We recently checked that there is no performance penalty when using LINQ for this:
start = oldSnapshot.IssueMarkers.First().Span....
end = oldSnapshot.IssueMarkers.Last().Span....

The same could be applied to the code below.